Spark: add spark extension for identifier fields

jackye1995 opened the pull request:
continuation of #2556 , will need to rebase after that is merged.

Here I propose the following syntax:

```sql
ALTER TABLE table REPLACE IDENTIFIER FIELDS (field1, field2, ...)

-- drop all
ALTER TABLE table DROP IDENTIFIER FIELDS
```

`IDENTIFIER` and `FIELDS` are new keywords. Because we already have `FIELD` as a keyword, I think adding `FIELDS` can be beneficial for other batch operations in the future. It also makes the statement more clear instead of saying `ALTER TABLE table REPLACE IDENTIFIER`.

The reason for having another drop statement to drop all is because typically in SQL statements a `( ... )` clause always has at least one element. We can do `ALTER TABLE table REPLACE IDENTIFIER FIELDS ()` instead, but it just feels awkward to me, not sure how other people think about this.

jackye1995 commented on api/src/main/java/org/apache/iceberg/types/TypeUtil.java:
```diff
@@ -22,10 +22,12 @@
 import java.util.List;
 import java.util.Locale;
```
contents in this file will be rebased out.

jackye1995 commented on spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:
```diff
@@ -159,7 +161,7 @@ quotedIdentifier
 nonReserved
     : ADD | ALTER | AS | ASC | BY | CALL | DESC | DROP | FIELD | FIRST | LAST | NULLS | ORDERED | PARTITION | TABLE | WRITE
-    | DISTRIBUTED | LOCALLY | UNORDERED
+    | DISTRIBUTED | LOCALLY | UNORDERED | REPLACE | WITH | ROW_IDENTIFIER | FIELDS
```
also added missing keywords that I forgot to add in #2365

jackye1995 commented:
@openinx @rdblue @yyanyy could you take a look? Thanks!

yyanyy commented on spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:
```diff
@@ -71,6 +71,8 @@ statement
     | ALTER TABLE multipartIdentifier DROP PARTITION FIELD transform                        #dropPartitionField
     | ALTER TABLE multipartIdentifier REPLACE PARTITION FIELD transform WITH transform (AS name=identifier)? #replacePartitionField
     | ALTER TABLE multipartIdentifier WRITE writeSpec                                       #setWriteDistributionAndOrdering
+    | ALTER TABLE multipartIdentifier REPLACE ROW_IDENTIFIER FIELDS '(' fields+=identifier (',' fields+=identifier)* ')' #replaceIdentifierFields
```
do we now have `SET ROW_IDENTIFIER`? if not, wondering if it would seem odd if user can only "replace" even when the table doesn't have row identifiers yet

openinx replied:
+1 for @yyanyy 's comment.

jackye1995 replied:
yeah `SET` makes sense to me. Was trying to reuse `REPLACE` keyword, but it does feel awkward.

yyanyy commented on spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/ReplaceIdentifierFieldsExec.scala:
```diff
@@ -0,0 +1,52 @@
+    catalog.loadTable(ident) match {
+      case iceberg: SparkTable =>
+        iceberg.table.updateSchema()
+          .setIdentifierFields(scala.collection.JavaConverters.seqAsJavaList(fields))
```
not familiar with this code base, curious - is it possible that input `fields` is null here and we encounter NPE? If it's nullable and NPE safe, do we want `replace` with empty input to be the same as `drop`?

yyanyy commented on spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestAlterTableSchema.java:
```diff
@@ -0,0 +1,85 @@
+import org.junit.Assert;
+import org.junit.Test;
+
+public class TestAlterTableSchema extends SparkExtensionsTestBase {
```
great to see this!

openinx commented:
I would like to invite @aokolnychyi to review this patch !

jackye1995 commented:
@yyanyy I changed to use `SET` and `UNSET` instead of `REPLACE` and `DROP`, please see it that feels better, thanks

rdblue commented on spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:
```diff
@@ -71,6 +71,8 @@ statement
     | ALTER TABLE multipartIdentifier DROP PARTITION FIELD transform                        #dropPartitionField
     | ALTER TABLE multipartIdentifier REPLACE PARTITION FIELD transform WITH transform (AS name=identifier)? #replacePartitionField
     | ALTER TABLE multipartIdentifier WRITE writeSpec                                       #setWriteDistributionAndOrdering
+    | ALTER TABLE multipartIdentifier SET ROW_IDENTIFIER FIELDS '(' fields+=identifier (',' fields+=identifier)* ')' #setIdentifierFields
```
Should `fields` be a list of `multipartIdentifier` since we support nested fields?

rdblue commented on spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:
```diff
@@ -159,7 +161,7 @@ quotedIdentifier
 nonReserved
     : ADD | ALTER | AS | ASC | BY | CALL | DESC | DROP | FIELD | FIRST | LAST | NULLS | ORDERED | PARTITION | TABLE | WRITE
-    | DISTRIBUTED | LOCALLY | UNORDERED
+    | DISTRIBUTED | LOCALLY | UNORDERED | REPLACE | WITH | ROW_IDENTIFIER | FIELDS | SET | UNSET
```
Looks like `REPLACE` is here, but should not be?

jackye1995 replied:
we forgot to add REPLACE in an older PR, so I added it here.

rdblue commented on spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:
```diff
@@ -174,15 +176,19 @@ DESC: 'DESC';
 ORDERED: 'ORDERED';
 PARTITION: 'PARTITION';
 REPLACE: 'REPLACE';
+ROW_IDENTIFIER: 'IDENTIFIER';
```
What about naming this `IDENTIFIER_KW` or something so it is more clear that this is `identifier` as a keyword?

rdblue commented on spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:
```diff
@@ -71,6 +71,8 @@ statement
     | ALTER TABLE multipartIdentifier REPLACE PARTITION FIELD transform WITH transform (AS name=identifier)? #replacePartitionField
     | ALTER TABLE multipartIdentifier WRITE writeSpec                                       #setWriteDistributionAndOrdering
+    | ALTER TABLE multipartIdentifier SET ROW_IDENTIFIER FIELDS '(' fields+=identifier (',' fields+=identifier)* ')' #setIdentifierFields
+    | ALTER TABLE multipartIdentifier UNSET ROW_IDENTIFIER FIELDS '(' fields+=identifier (',' fields+=identifier)* ')' #unsetIdentifierFields
```
What is the behavior of `UNSET IDENTIFIER FIELDS`? That isn't obvious to me so I'm wondering if we should change it.

rdblue replied:
What about `DROP` instead of `UNSET` since this is dropping identifier fields from the set?

rdblue commented on spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/UnsetIdentifierFieldsExec.scala:
```diff
@@ -0,0 +1,55 @@
+  override protected def run(): Seq[InternalRow] = {
+    catalog.loadTable(ident) match {
+      case iceberg: SparkTable =>
+        val identifierFieldNames = iceberg.table.schema().identifierFieldNames()
```
Nit: it's better to omit parens with Java APIs when they would be omitted in a Scala API. Since this is a getter, I would omit the parens on `schema` and `identifierFieldNames`.

rdblue commented on spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/UnsetIdentifierFieldsExec.scala:
```diff
@@ -0,0 +1,55 @@
+    catalog.loadTable(ident) match {
+      case iceberg: SparkTable =>
+        val identifierFieldNames = iceberg.table.schema().identifierFieldNames()
+        fields.map(f => identifierFieldNames.remove(f))
```
This should not assume that it is safe to modify `identifierFieldNames`. The class is free to return an immutable collection, and arguably should return one. We don't want to rely on the existing behavior.

jackye1995 commented:
@rdblue updated based on the suggestions to use `DROP`. Also fixed a bug along the way regarding nested field in `Schema`.

rdblue commented on api/src/main/java/org/apache/iceberg/Schema.java:
```diff
@@ -215,7 +215,7 @@ public StructType asStruct() {
     return identifierFieldIds()
             .stream()
-            .map(id -> findField(id).name())
+            .map(id -> lazyIdToName().get(id))
```
Good catch. It took me a minute to realize what the difference was.

rdblue commented on spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DropIdentifierFieldsExec.scala:
```diff
@@ -0,0 +1,55 @@
+    catalog.loadTable(ident) match {
+      case iceberg: SparkTable =>
+        val identifierFieldNames = iceberg.table.schema().identifierFieldNames()
+        fields.map(f => identifierFieldNames.remove(f))
```
@jackye1995, looks like you didn't see my earlier comment:

This should not assume that it is safe to modify `identifierFieldNames`. The `Schema` class is free to return an immutable collection, and arguably _should_ return one. We don't want to rely on the existing behavior that it is modifiable.

Can you copy the identifier field names collection and modify the copy instead?

I think this should also validate that all of the fields listed in the `DROP` call are actually identifier fields. It would be an error if I ran `DROP IDENTIFIER FIELDS (manager_id)` but the rows are currently identified by `user_id`.

rdblue commented on spark3-extensions/src/main/antlr/org.apache.spark.sql.catalyst.parser.extensions/IcebergSqlExtensions.g4:
```diff
@@ -71,6 +71,8 @@ statement
     | ALTER TABLE multipartIdentifier DROP PARTITION FIELD transform                        #dropPartitionField
     | ALTER TABLE multipartIdentifier REPLACE PARTITION FIELD transform WITH transform (AS name=identifier)? #replacePartitionField
     | ALTER TABLE multipartIdentifier WRITE writeSpec                                       #setWriteDistributionAndOrdering
+    | ALTER TABLE table=multipartIdentifier SET IDENTIFIER_KW FIELDS '(' fields+=multipartIdentifier (',' fields+=multipartIdentifier)* ')' #setIdentifierFields
```
You might consider allowing more flexible syntax here, like `SET IDENTIFIER FIELD user_id` since `FIELD` will make more sense to users. Should the parens be optional?

Could you move the `fields` to a `fieldList` rule so that we don't have two copies of it?

jackye1995 replied:
I think we have the following options to choose from:
- ADD IDENTIFIER FIELD
- ADD IDENTIFIER FIELDS
- SET IDENTIFIER FIELD
- SET IDENTIFIER FIELDS
- DROP IDENTIFIER FIELD
- DROP IDENTIFIER FIELDS

I am choosing to implement the batch operations because that allows one single commit to add or drop multiple fields. Personally I don't feel it is hard for users to specify `SET IDENTIFIER FIELDS (user_id)`, given that all Hive DDLs are in this batch format. Do you think it is worth having a clause also for single field modifications in this case?

If we want to have a way to only add new fields without the need to know current fields, I can also add a `ADD IDENTIFIER FIELDS`.

rdblue replied:
My comment was based on similar reasoning for other DDL. For example, it is better to support both `ADD COLUMNS` and `ADD COLUMN` interchangeably so that users don't get needless syntax errors when they use a plural or singular term with the "wrong" number of columns.

I think that the problem extending that to this is that it isn't clear that `SET IDENTIFIER FIELD x` actually replaces all existing identifier fields, which I think is why you're mentioning not wanting to use the `ADD IDENTIFIER FIELD` syntax. I think that if there is confusion now, then there would be more confusion for users, so maybe it is better to only support `SET IDENTIFIER FIELDS (...)`. Though I would make the parens optional.

> . . . given that all Hive DDLs are in this batch format

I don't think that we want to take inspiration from Hive very much. Better to use Postgres or Trino as reference points.

> If we want to have a way to only add new fields without the need to know current fields, I can also add a ADD IDENTIFIER FIELDS.

This wasn't my intent and I would not want to have confusion, so let's not allow using `FIELD` interchangeably.

jackye1995 replied:
> I would make the parens optional

Do you mean to also not have parens optional when there are multiple fields, like:

```
SET IDENTIFIER FIELDS first_name, last_name
```

Or is it only for the case when there is 1 field?

rdblue replied:
Yeah, I would make it possible to use that syntax. No need for users to remember that they need parens when we don't actually need them to parse correctly.

jackye1995 replied:
Okay that makes sense. I will update with all the other comments, thanks!

rdblue commented on spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestAlterTableSchema.java:
```diff
@@ -0,0 +1,100 @@
+  }
+
+  @Test
+  public void testDropIdentifierFields() {
```
It would be good to have a test for dropping an identifier field that isn't an identifier field, and for dropping a field that isn't a table field. The error messages should probably be different for those cases?

rdblue commented on spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestAlterTableSchema.java:
```diff
@@ -0,0 +1,100 @@
+  }
+
+  @Test
+  public void testSetIdentifierFields() {
```
It would be good to have a test for using an identifier field that is not defined in the table.

jackye1995 commented:
@rdblue sorry I was stuck with some internal work and did not get time to update this. the latest code addressed most of the comments.

I have been thinking about the syntax over the last few days, I think I was very biased with the Hive DDL syntax, but as you pointed out, if we prefer the other way, it might be better to instead use a syntax like:

```sql
ALTER TABLE table WRITE IDENTIFIED BY first_name,last_name;
```

And we don't support any other syntax like add and drop.

This avoids the confusion and it should be clear that we are replacing all the identifier fields. It is closer to the existing syntax of sort order, plus it does not need parentheses. What do you think?

rdblue commented on spark3-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DropIdentifierFieldsExec.scala:
```diff
@@ -0,0 +1,65 @@
+
+        for (name <- fields) {
+          Preconditions.checkArgument(schema.findField(name) != null,
+            "Cannot complete drop identifier fields operation: field %s not found", name)
```
Nit: you can omit "complete" and "operation" to be more concise. "Cannot drop identifier field: %s (not found)" says basically the same thing.

rdblue commented:
@jackye1995, sorry, I didn't read your last comment before looking at the updates. I approved and merged this.

I like your idea to change this to `WRITE IDENTIFIED BY`. I'd like to hear what @aokolnychyi thinks about it. I'm happy with the current syntax as well.

jackye1995 commented:
Okay let me create an issue to discuss that
